pandas/core/resample.py

```diff
@@ -1269,6 +1239,58 @@ def _get_time_period_bins(self, ax):
 
         return binner, bins, labels
 
+    def _get_period_bins(self, ax):
+        if not isinstance(ax, PeriodIndex):
+            raise TypeError('axis must be a PeriodIndex, but got '
+                            'an instance of %r' % type(ax).__name__)
+
+        memb = ax.asfreq(self.freq, how=self.convention)
+        # NaT handling as in pandas._lib.lib.generate_bins_dt64()
+        nat_count = 0
+        if memb.hasnans:
+            import warnings
+            with warnings.catch_warnings():
+                warnings.filterwarnings('ignore', 'numpy equal will not check '
+                                                  'object identity')
+                nat_mask = memb.base == tslib.NaT
+                # raises "FutureWarning: numpy equal will not check object
+                # identity in the future. The comparison did not return the
+                # same result as suggested by the identity (`is`)) and will
+                # change."
+            nat_count = np.sum(nat_mask)
+            memb = memb[~nat_mask]
+
+        # if index contains no valid (non-NaT) values, return empty index
+        if not len(memb):
+            binner = labels = PeriodIndex(
+                data=[], freq=self.freq, name=ax.name)
+            return binner, [], labels
+
+        start = ax.min().asfreq(self.freq, how=self.convention)
+        end = ax.max().asfreq(self.freq, how='end')
+
+        labels = binner = PeriodIndex(start=start, end=end,
+                                      freq=self.freq, name=ax.name)
+
+        i8 = memb.asi8
+        freq_mult = self.freq.n
+        # when upsampling to subperiods, we need to generate enough bins
+        expected_bins_count = len(binner) * freq_mult
+        i8_extend = expected_bins_count - (i8[-1] - i8[0])
```

Comment:
Some lines above (after masking out NaTs), we check for ```if not len(memb)``` and return and empty index, so this case should be caught.
Maybe put in an ```assert len(memb)``` to make it explicit?